File: codechef/dec14/xorsub.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.StringTokenizer;

class Main
{
	public static int calculateLength(int n) throws Exception
	{
		if(n<=0)
			return 1;
		else
			return (int)(Math.log(n)/Math.log(2))+1;
	}
	public static void main(String args[])
	{
		try
		{
			BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
			int t=Integer.parseInt(br.readLine());
			while(t-->0)
			{
				StringTokenizer stz=new StringTokenizer(br.readLine());
				int n=Integer.parseInt(stz.nextToken());
				int k=Integer.parseInt(stz.nextToken());
				int arr[]=new int[n];
				stz=new StringTokenizer(br.readLine());
				for(int i=0;i<n;i++)
					arr[i]=Integer.parseInt(stz.nextToken());
				LinkedList list[]=new LinkedList[11];
				for(int i=0;i<list.length;i++)
					list[i]=new LinkedList<Integer>();
				for(int i=0;i<n;i++)
					list[calculateLength(arr[i])].add(new Integer(arr[i]));
				// Gaussian Elimination
				LinkedList newList[]=new LinkedList[11];
				for(int i=0;i<newList.length;i++)
					newList[i]=new LinkedList<Integer>();
				for(int i=10;i>1;i--)
				{
					if(!(list[i].isEmpty()))
					{
						int xelement=(Integer)list[i].get(0);
						newList[i].add(new Integer(xelement));
						list[i].removeFirst();
						int element=((Integer)(newList[i].get(0))).intValue();
						while(!list[i].isEmpty())
						{
							int oldElement=((Integer)list[i].get(0)).intValue();
							list[i].removeFirst();
							int len=calculateLength(element^oldElement);
							if(len>0)
								list[len].add(new Integer(element^oldElement));
						}
					}
				}
				int x=k;
				if(!(list[1].isEmpty()))
					newList[1].add((Integer)(list[1].get(0)));
				for(int i=10;i>=1;i--)
				{
					//System.out.println("i="+i+" x="+x);
					if(!newList[i].isEmpty())
					{
						int element=((Integer)newList[i].get(0)).intValue();
						if((x^element) > x)
							x=(x^element);
					}
				}
				System.out.println(x);
			}
		}
		catch(Exception e)
		{

		}
	}
}

File: codechef/dec14/xorsub.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
struct Node
{
    int data;
    struct Node*next;
};

struct Node* makeNode(int n)
{
    struct Node*node;
    node=(struct Node*)malloc(sizeof(struct Node)*n);
    node->next=NULL;
    node->data=n;
    return node;
}

void printList(struct Node*start)
{
    while(start)
    {
        printf("%d\n",start->data);
        start=start->next;
    }
}

struct Node*add(struct Node*start,struct Node*newNode)
{
    struct Node*node;
    if(!start)
        return newNode;
    for(node=start;node->next;node=node->next);
    node->next=newNode;
    return start;
}
int calculateLength(int n)
{
    int answer=0;
    if(n<=0)
        return 0;

    while(n)
    {
        answer++;
        n>>=1;
    }
    return answer;
}
struct Node*removeFirst(struct Node*start)
{
    struct Node*temp;
    if(!start)
        return start;
    temp=start;
    start=start->next;
    free(temp);
    return start;
}
int main()
{
    int t,n,k,*arr,a,i,x,xelement,oldElement,len,element;
    struct Node *list[11];
    struct Node *newList[11];
    scanf("%d",&t);
    while(t--)
    {
        for(i=0;i<=10;i++)
        {
            list[i]=NULL;
            newList[i]=NULL;
        }
        scanf("%d%d",&n,&k);
        arr=(int*)malloc(sizeof(int)*n);
        for(i=0;i<n;i++)
            scanf("%d",&arr[i]);
        for(i=0;i<n;i++)
        {
            a=calculateLength(arr[i]);
            list[a]=add(list[a],makeNode(arr[i]));
        }
        // Gaussian Elimination
        for(i=10;i>=0;i--)
        {
            //printf("hello lavish kothari\n");
            if(list[i])
            {
                xelement=list[i]->data;
                newList[i]=add(newList[i],makeNode(xelement));
                list[i]=removeFirst(list[i]);
                element=newList[i]->data;
                while(list[i])
                {
                    oldElement=list[i]->data;
                    list[i]=removeFirst(list[i]);
                    len=calculateLength(element^oldElement);
                    if(len>0)
                        list[len]=add(list[len],makeNode(element^oldElement));
                }
            }
        }
        x=k;

        for(i=10;i>=1;i--)
        {
            if(newList[i])
            {
                element=newList[i]->data;
                if((x^element) > x)
                    x=(x^element);
            }
        }
        printf("%d\n",x);
        for(i=0;i<=10;i++)
        {
            if(list[i])
                free(list[i]);
            if(newList[i])
                free(newList[i]);
        }
    }
    return 0;
}
